proto/osmosis/gamm/v1beta1/pool.proto

```diff
@@ -32,6 +35,57 @@ message PoolAsset {
   // ];
 }
 
+// Parameters for changing the weights in a balancer pool smoothly from
+// a start weight and end weight over a period of time.
+// Currently, the only smooth change supported is linear changing between
+// the two weights, but more types may be added in the future.
+// When these parameters are set, the weight w(t) for pool time `t` is the following:
+//   t <= start_time: w(t) = initial_pool_weights
+//   start_time < t <= start_time + duration: 
+//     w(t) = initial_pool_weights + (t - start_time) * 
+//       (target_pool_weights - initial_pool_weights) / (duration)
+//   t > start_time + duration: w(t) = target_pool_weights
+message SmoothWeightChangeParams {
+  // The start time for beginning the weight change.
+  // If a parameter change / pool instantiation leaves this blank,
+  // it should be generated by the state_machine as the current time.
+  google.protobuf.Timestamp start_time = 1 [
+    (gogoproto.stdtime) = true,
+    (gogoproto.nullable) = false,
+    (gogoproto.moretags) = "yaml:\"start_time\""
+  ];
+  // Duration for the weights to change over
+  google.protobuf.Duration duration = 2 [
+    (gogoproto.nullable) = false,
+    (gogoproto.stdduration) = true,
+    (gogoproto.jsontag) = "duration,omitempty",
+    (gogoproto.moretags) = "yaml:\"duration\""
+  ];
+  // The initial pool weights. These are copied from the pool's settings
+  // at the time of weight change instantiation.
+  // The amount PoolAsset.token.amount field is ignored if present,
+  // future type refactorings should just have a type with the denom & weight here.
+  repeated PoolAsset initialPoolWeights = 3 [
+    (gogoproto.moretags) = "yaml:\"initial_pool_weights\"",
+    (gogoproto.nullable) = false
+  ];
+  // The target pool weights. The pool weights will change linearly with respect to time
+  // between start_time, and start_time + duration. 
+  // The amount PoolAsset.token.amount field is ignored if present,
+  // future type refactorings should just have a type with the denom & weight here.
+  repeated PoolAsset targetPoolWeights = 4 [
+    (gogoproto.moretags) = "yaml:\"target_pool_weights\"",
+    (gogoproto.nullable) = false
+  ];
+  // Intermediate variable for the 'slope' of pool weights. This is equal to 
```

Comment:
We can at least cache the `totalWeightsDiff` so that doesn't have to be recalculated each time

`subPoolAssetWeights(params.TargetPoolWeights, params.InitialPoolWeights)` shouldn't change

Reply:
Agreed, though its of unclear merit, since that just increases state reads, for negligible computational benefit. (Its just saving a subtraction)

Reply:
Should be same state reads I think, because you don't have to query TargetPoolWeights, but yeah pretty negligible.

Reply:
Its the same number of reads, but not the same amount of data